archipelago/pipeline.py

```diff
@@ -0,0 +1,1028 @@
+                elif g_node.route_type == "RMUX":
+                    if graph.get_node(parent).route_type != "REG":
+                        comp.available_regs += 1
+                if graph.node_latencies[node] != 0:
+                    comp.glbs = 0
+                    comp.hhops = 0
+                    comp.uhops = 0
+                    comp.dhops = 0
+                    comp.pes = 0
+                    comp.mems = 0
+                    comp.available_regs = 0
+                    comp.parent = None
+            components.append(comp)
+        
+        maxt = 0
+        max_comp = components[0]
+        for comp in components:
+            if comp.get_total() > maxt:
+                maxt = comp.get_total()
+                max_comp = comp
+
+        timing_info[node] = max_comp
+
+    node_to_timing = {node:timing_info[node].get_total() for node in graph.nodes}
+    node_to_timing = dict(sorted(reversed(list(node_to_timing.items())), key=lambda item: item[1], reverse=True))
+    max_node = list(node_to_timing.keys())[0]
+    max_delay = list(node_to_timing.values())[0]
+
+
+    clock_speed = 1.0e12 / max_delay / 1e6
+
+    print("\nCritical Path Info:")
+    print("\tMaximum clock frequency:", clock_speed, "MHz")
+    print("\tCritical Path:", max_delay, "ns")
+
+    net_ids = set()
+
+    curr_node = max_node
+    net_ids.add(graph.get_node(curr_node).net_id)
+    print(f"\t{curr_node}", "glb:", timing_info[curr_node].glbs, "horiz hops:",  timing_info[curr_node].hhops, "up hops:",  timing_info[curr_node].uhops, "down hops:",  timing_info[curr_node].dhops, "pes:", timing_info[curr_node].pes, "mems:", timing_info[curr_node].mems, "\n")
+    crit_path = []
+    crit_path.append((curr_node, timing_info[curr_node].get_total()))
+    crit_edges = []
+    while(True):
+        if (timing_info[curr_node].parent, curr_node) in graph.edges:
+            crit_edges.append((timing_info[curr_node].parent, curr_node))
+        curr_node = timing_info[curr_node].parent
+        if curr_node == None:
+            break
+        net_ids.add(graph.get_node(curr_node).net_id)
+        crit_path.append((curr_node, timing_info[curr_node].get_total()))
+
+    crit_path.reverse()
+
+    #print("\tCritical Path Nets:", *net_ids)
+
+    return clock_speed, crit_path, crit_edges
+
+def find_break_idx(graph, crit_path):
+    crit_path_adjusted = [abs(c - crit_path[-1][1]/2) for n,c in crit_path]
+    break_idx = crit_path_adjusted.index(min(crit_path_adjusted))
+
+    if len(crit_path) < 2:
+        raise ValueError("Can't find available register on critical path")
+
+    while True:
+        if graph.get_node(crit_path[break_idx + 1][0]).route_type == "RMUX" and graph.get_node(crit_path[break_idx][0]).route_type == "SB":
+            return break_idx
+        break_idx += 1
+
+        if break_idx + 1 >= len(crit_path):
+            break_idx = crit_path_adjusted.index(min(crit_path_adjusted))
+
+            while True:

```

Comment:
I'm not sure I'm following the logic. are these two level of while true loop necessary? can't you just use two separate for loops? Nested loop structures should be avoided if possible since it decrease the code readability. Same thing goes for other logic in the PR.